src/typechecker.erl

```diff
@@ -2634,11 +2638,28 @@ type_check_comprehension(Env, Compr, Expr, [Guard | Quals]) ->
       ResTy :: type(),
       Expr :: expr().
 type_check_expr_in(Env, ResTy, Expr) ->
+    {InferredTy, NewEnv} = type_check_expr(Env, Expr),
     ?verbose(Env, "~sChecking that ~ts :: ~ts~n",
-            [gradualizer_fmt:format_location(Expr, brief), erl_prettypr:format(Expr), typelib:pp_type(ResTy)]),
-    NormResTy = normalize(ResTy, Env),
-    R = ?throw_orig_type(do_type_check_expr_in(Env, NormResTy, Expr), ResTy, NormResTy),
-    ?assert_type(R, env()).
+            [gradualizer_fmt:format_location(Expr, brief),
+             erl_prettypr:format(Expr),
+             typelib:pp_type(ResTy)]),
+    case subtype(InferredTy, ResTy, NewEnv) of
+        true ->
+            NewEnv;
+        false ->
+            throw(type_error(Expr, InferredTy, ResTy))
+    end.
```

Comment:
The state of the art in writing typecheckers is to do it the other way around, compared to what you've done here. Always pass in the type that's expected, though if we don't know the type just pass a variable. You'll likely get better propagation of types that way.

See e.g. slide 43 in this presentation by Simon Peyton Jones: https://haskell.foundation/assets/other/Simon%20Peyton%20Jones%20-%202023-06-08%20-%20Type%20inference%20in%20GHC%20Jun%2023.pdf
Video can be found here: https://haskell.foundation/events/2023-ghc-development-workshop.html